### PR TITLE
🦺 #285: improve optional regex validation

### PR DESCRIPTION
**Pull request by samrichca**
### Consent UI

- Create separate schema for the Clinician Invite form that extends the ClinicianInviteBase schema to use `EmptyOrOptionalName` for the PreferredName field

### Types

- Add utilities for string validations in Zod
- Add tests for nano ID
- Update regex validation for Name, Phone Number, NanoId, OhipNumber, and PostalCode
  - Add validation for optional fields
  - Add validation for empty or optional fields, for form validation (not for NanoId because users will never enter it in a form)
- Move refinements from base schemas to response/request schemas 

## Instructions

- In the packages/types folder, run `pnpm run build`
- Then in the root folder, run `pnpm run build` again
- Run `pnpm run test` to see all the tests

## PR Readiness Checklist

- [x] "Expected Outcome(s)" in ticket have been met
- [x] Ticket number included in PR title
- [x] Connected ticket to PR
- [x] Labels added to PR for service name (`consent-api`, `data-mapper`, etc...), type (`chore`, `documentation`, etc...), status (`draft`, `on-hold`, etc...) **if applicable**
- [x] Manual testing completed
- [x] Builds locally without errors or warnings
- [x] Tests are updated (if required) and passing
- [ ] PR feedback has been addressed **if applicable**
- [x] New environment variables added to `.env.schema` files, `README.md`
- [x] Added copyrights to new files

### CODE REVIEW

File: packages/types/src/entities/ClinicianInvite.ts [hunk 1]
```diff
@@ -60,9 +60,10 @@ export const InviteEntity = z.object({
 	inviteAccepted: z.boolean().default(false),
 });
 
+export const InviteFieldsPreRefine =
+	InviteClinicianFields.merge(InviteGuardianFields).merge(InviteParticipantFields);
+
 // base type req/res for the "full" invite types used in consent-api and data-mapper
-export const ClinicianInviteBase = InviteClinicianFields.merge(InviteGuardianFields)
-	.merge(InviteParticipantFields)
-	.refine(hasRequiredGuardianInformation, {
```
Comment (samrichca) [inline packages/types/src/entities/ClinicianInvite.ts:66]: we can't chain `merge()` after `refine()` so I had to split up these schemas.

File: apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx [hunk 2]
```diff
@@ -52,6 +54,13 @@ import formStyles from './ClinicianInviteForm.module.scss';
 
 const styles = Object.assign({}, formStyles, layoutStyles);
 
+const ClinicianInviteUI = InviteFieldsPreRefine.extend({
+	participantPreferredName: NameOptionalUI,
+}).refine(hasRequiredGuardianInformation, {
+	message: 'Guardian contact fields are required for that consentGroup',
```
Comment (samrichca) [inline apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx:60]: I think we're going to need different error handling in the UI for custom errors like this, because of translation, so I repeated this error message here instead of abstracting it into a reusable function.

File: packages/types/tests/entities/Name.spec.ts [hunk 3]
```diff
@@ -36,3 +36,35 @@ describe('Name', () => {
 		expect(Name.safeParse(' ').success).false;
 	});
 });
+
+describe('NameOptionalAPI', () => {
```
Comment (samrichca) [inline packages/types/tests/entities/Name.spec.ts:40]: I know we said don't test the Zod API but I think testing regexes is a good call.

File: packages/types/src/entities/ClinicianInvite.ts (hunk 1 above)
Comment (samrichca) [reply packages/types/src/entities/ClinicianInvite.ts:66]: tried to use `intersection()` with the original code, i.e. `z.intersection(ClinicianInviteBase, z.object({ participantPreferredName: NameOptionalUI }))` and it didn't work. I still got an error when the field was empty.

**Review by anncatton — COMMENTED** [review]
I've added a [supplemental PR](https://github.com/OHCRN/platform/pull/445) off of this one, with my suggested approach. Please take a look and try it out, let me know how it looks to you.

I'll tag @mistryrn and @xusandrew on that one as well.

File: docs/shared-types.md [hunk 4]
```diff
@@ -8,14 +8,14 @@ This document describes how we use [Zod](https://zod.dev/) schemas to validate a
...
-| Base schemas | [`types/src/entities`](../packages/types/src/entities/) | | `'types/entities'` |
-| Field types | [`types/src/entities/fields`](../packages/types/src/entities/fields/) | types for fields in the data model, such as enums | `'types/entities'` |
-| Request schemas | [`types/src/services/<service-name>/requests`](../packages/types/src/services/<service-name>/requests/) | Schemas extended from applicable base type, used for requests to the specified service | `'types/<service-name>` |
-| Response schemas | [`types/src/services/<service-name>/responses`](../packages/types/src/services/<service-name>/responses/) | Schemas extended from applicable base type, used for responses from the specified service | `'types/<service-name>` |
-| HTTP response schemas | [`types/src/httpResponses`](../packages/types/src/httpResponses/) | Generic HTTP success and error responses for all services |
+| Category              | Directory                                                                                                 | Description                                                                                   | Exported as             |
```
Comment (samrichca) [inline docs/shared-types.md:11]: formatting from prettier

File: apps/consent-ui/src/components/views/ConsentWizard/ConsentReviewSign/ConsentReviewSignForm.tsx [hunk 5]
```diff
@@ -31,7 +32,7 @@ import { useNotification } from 'src/components/providers/NotificationProvider';
 import useGoToNextConsentStep from '../ConsentStepsNavigation/useGoToNextConsentStep';
 import ConsentStepsNavigation from '../ConsentStepsNavigation';
 
-export const ConsentReviewSignRequest = z.object({ stub: z.string().min(1) });
+export const ConsentReviewSignRequest = z.object({ stub: RequiredString });
```
Comment (samrichca) [inline apps/consent-ui/src/components/views/ConsentWizard/ConsentReviewSign/ConsentReviewSignForm.tsx:35]: this changed to `z.string().trim().min(1)` to validate against whitespace.

File: packages/types/src/entities/fields/NanoId.ts [hunk 6]
```diff
@@ -26,6 +26,11 @@ import { NANOID_REGEX } from '../../common/regexes.js';
 export const ID_ALPHABET =
 	'0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz' as const;
 
+// string with regex
 export const NanoId = z.string().regex(NANOID_REGEX);
```
Comment (samrichca) [inline packages/types/src/entities/fields/NanoId.ts:30]: nano IDs are never user generated so we don't need to trim, or accept whitespace.

File: packages/types/src/entities/fields/OhipNumber.ts [hunk 7]
```diff
@@ -19,7 +19,20 @@
...
+// string with regex
 export const OhipNumber = z.string().regex(OHIP_NUMBER_REGEX);
 export type OhipNumber = z.infer<typeof OhipNumber>;
+
+// optional string with regex
+export const OptionalOhipNumber = OhipNumber.optional();
+export type OptionalOhipNumber = z.infer<typeof OptionalOhipNumber>;
+
+// optional trimmed string with regex OR empty/whitespace string
+export const EmptyOrOptionalOhipNumber = TrimmedString.regex(OHIP_NUMBER_REGEX)
```
Comment (samrichca) [inline packages/types/src/entities/fields/OhipNumber.ts:34]: I've opted to only use trimmed strings for 'EmptyOrOptional' because OhipNumber & OptionalOhipNumber should be validating pre-trimmed strings.

File: packages/types/src/entities/fields/PhoneNumber.ts [hunk 8]
```diff
@@ -19,7 +19,20 @@
...
+// string with regex
+export const PhoneNumber = z.string().regex(PHONE_NUMBER_REGEX);
 export type PhoneNumber = z.infer<typeof PhoneNumber>;
+
+// optional string with regex
+export const OptionalPhoneNumber = PhoneNumber.optional();
+export type OptionalPhoneNumber = z.infer<typeof OptionalPhoneNumber>;
+
+// optional trimmed string with regex OR empty/whitespace string
+export const EmptyOrOptionalPhoneNumber = TrimmedString.regex(PHONE_NUMBER_REGEX)
```
Comment (samrichca) [inline packages/types/src/entities/fields/PhoneNumber.ts:34]: same here as OHIP number. I only used trimmed strings for validating user input.

File: packages/types/src/services/consentApi/requests/ClinicianInvite.ts [hunk 9]
```diff
@@ -21,8 +21,9 @@ import { generateSchema } from '@anatine/zod-openapi';
 import type { SchemaObject } from 'openapi3-ts/oas31';
 import { z } from 'zod';
 
+import { hasRequiredGuardianInformation } from '../../../common/index.js';
 import { ClinicianInviteBase } from '../../../entities/ClinicianInvite.js';
 
-export const ClinicianInviteRequest = ClinicianInviteBase;
+export const ClinicianInviteRequest = ClinicianInviteBase.refine(hasRequiredGuardianInformation);
```
Comment (samrichca) [inline packages/types/src/services/consentApi/requests/ClinicianInvite.ts:27]: moved refinements to request/response rather than base (thanks @anncatton )

**Review by anncatton — COMMENTED** [review]
Good stuff here, these more primitive types are very readable! Will be glad to get this into to develop so I can pick up some of these changes my current DAS prs.

Left a few recos for cleanup/ documentation, and a few questions for clarification. I may be missing some context on the Ohip and PhoneNumber types.

**Comment by samrichca** [comment]
"Also curious, where will EmptyOrOptionalPhoneNumber be used?" I was thinking about optional PhoneNumber.optional() fields, but in the UI those are required or hidden, so they can't be empty strings. I'll remove this one

File: apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx [hunk 10]
```diff
@@ -52,6 +54,12 @@ import formStyles from './ClinicianInviteForm.module.scss';
 
 const styles = Object.assign({}, formStyles, layoutStyles);
 
+const ClinicianInviteUI = ClinicianInviteBase.extend({
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx:57]: it's only being used for form validation, would `ClinicianInviteFormValidation` work?

File: packages/types/src/entities/fields/PhoneNumber.ts (hunk 8 above)
Comment (samrichca) [reply packages/types/src/entities/fields/PhoneNumber.ts:34]: we have some optional phone numbers but on second thought, those are conditionally required, so in the UI they have to be phone numbers or undefined, they can't be empty/whitespace. I'll go ahead and remove EmptyOrOptional

File: packages/types/src/entities/fields/OhipNumber.ts (hunk 7 above)
Comment (samrichca) [reply packages/types/src/entities/fields/OhipNumber.ts:34]: added some more tests to check this - if I use trimmed strings, then a phone number or postal code with trailing whitespace is valid, because the value being tested by the regex is a trimmed string. that's OK for the UI but not the API, I'd imagine.

image

**Review by anncatton — COMMENTED** [review]
Thanks for adding tests to verify how the optional regex fields are behaving. I just added a couple more questions/requests for clarification.

**Comment by samrichca** [comment]
"Just to make sure i understand the approach you're describing here: you want the UI form request types to use trimmed strings, but the schema the consent API (and any other backend service) uses will NOT do any string modifications, because it's going to assume that the data it's receiving is what has been intended to be sent."

yup! zod doesn't mutate data so backend services should take data at face value. before sending form data from the UI to API, we have to trim strings and convert empty strings to undefined.

**Comment by samrichca** [comment]
@anncatton "Just to make sure i understand the approach you're describing here: you want the UI form request types to use trimmed strings, but the schema the consent API (and any other backend service) uses will NOT do any string modifications, because it's going to assume that the data it's receiving is what has been intended to be sent."

yup! zod doesn't mutate data so validation in backend services should take data at face value and any data modifications should happen in a different operation. on the other hand, for UI, i have to check user input as it's happening. before sending form data from the UI to API, we have to trim strings (and convert empty strings to undefined) so i made the validation accept strings with whitespace.

File: docs/shared-types.md [hunk 11]
```diff
@@ -39,12 +39,43 @@ type ClinicianInvite = { ...baseFields };
...
+When values need to follow a specific pattern, such as 10 digits for a phone number, use Zod string and regular expression methods.
+
+For required fields: Use `z.string().regex(MY_REGEX)`. If the regex allows whitespace (such as `NAME_REGEX`), use `TrimmedString.regex()` to remove trailing whitespace and transform whitespace-only strings to empty strings. This is because some whitespace is allowed, but not strings that are only whitespace.
+
+```ts
+const Name = TrimmedString.regex(NAME_REGEX); // allows whitespace in regex
+const OhipNumber = z.string().regex(OHIP_NUMBER_REGEX); // doesn't allow whitespace in regex
+```
+
+For optional fields **in the API specifically**: Use the schema defined for the required field and chain `.optional()`. This will allow `undefined` values.
```
[the remark this answers is not in the document]
Comment (samrichca) [reply docs/shared-types.md]: updated

**Review by anncatton — COMMENTED** [review]
I think this is looking good to me. Really appreciate all the tests. Just want to clarify one more thing, your last comment:
```on the other hand, for UI, i have to check user input as it's happening. before sending form data from the UI to API, we have to trim strings (and convert empty strings to undefined) so i made the validation accept strings with whitespace.``` 
is referring to *actually modifying* the form data that's submitted (that's my assumption! 🙂 ) so that trailing whitespace is removed etc

File: packages/types/src/entities/Participant.ts [hunk 12]
```diff
@@ -57,7 +62,7 @@ export const ParticipantIdentityBase = ParticipantBaseOhipNameFields.merge(Parti
 	.merge(
 		z.object({
 			dateOfBirth: z.coerce.date(),
-			participantPreferredName: Name.optional(),
+			participantPreferredName: OptionalName,
 			keycloakId: z.string().uuid(),
 			inviteId: NanoId.optional(),
```
Comment (anncatton) [inline packages/types/src/entities/Participant.ts:67]: i missed this before - `inviteId` can use your `OptionalNanoId` type

File: packages/types/src/entities/Participant.ts [hunk 13]
```diff
@@ -67,12 +72,13 @@ export type ParticipantIdentityBase = z.infer<typeof ParticipantIdentityBase>;
...
-		mailingAddressStreet: z.string().optional(),
-		mailingAddressCity: z.string().optional(),
+		mailingAddressStreet: OptionalString,
+		mailingAddressCity: OptionalString,
 		mailingAddressProvince: Province.optional(),
-		mailingAddressPostalCode: PostalCode.optional(),
+		mailingAddressPostalCode: OptionalPostalCode,
 		residentialPostalCode: PostalCode,
-		participantOhipMiddleName: Name.optional(),
+		inviteId: OptionalNanoId,
```
Comment (anncatton) [inline packages/types/src/entities/Participant.ts:80]: can remove `inviteId` as it's already a part of `ParticipantIdentityBase`

**Review by anncatton — COMMENTED** [review]
Just noticed a couple smol things to update, i think this is from the earlier conflicts

**Review by anncatton — APPROVED** [review]
🚀 🚀 🐈
